Several performance and privacy improvements to inv/mempool handling

sipa opened the pull request:
As we're increasingly changing how block invs and tx invs are treated, bite the bullet and separate them into two separate queues. This makes sense, as they are subject to very different behaviour: we want blocks to relay as fast as possible, while transactions are subject to rate limiting, deduplication, and source obscurity.

Once this is done, we can turn the tx queue into a set, and use it to prevent the mempool command from leaking unannounced transactions. The handling of such commands is also moved to the send loop, sorted by dependency order, and made subject to trickling.

Based on top of  #7805.

gmaxwell committed:
Eliminate TX trickle bypass, sort TX invs for privacy and priority.

Previously Bitcoin would send 1/4 of transactions out to all peers
 instantly.  This causes high overhead because it makes >80% of
 INVs size 1.  Doing so harms privacy, because it limits the
 amount of source obscurity a transaction can receive.

These randomized broadcasts also disobeyed transaction dependencies
 and required use of the orphan pool.  Because the orphan pool is
 so small this leads to poor propagation for dependent transactions.

When the bypass wasn't in effect, transactions were sent in the
 order they were received.  This avoided creating orphans but
 undermines privacy fairly significantly.

This commit:
 Eliminates the bypass. The bypass is replaced by halving the
  average delay for outbound peers.

 Sorts candidate transactions for INV by their topological
  depth then by their feerate (then hash); removing the
  information leakage and providing priority service to
  higher fee transactions.

 Limits the amount of transactions sent in a single INV to
  7tx/sec (and twice that for outbound); this limits the
  harm of low fee transaction floods, gives faster relay
  service to higher fee transactions. The 7 sounds lower
  than it really is because received advertisements need
  not be sent, and because the aggregate rate is multipled
  by the number of peers.

sipa committed:
Split up and optimize transaction and block inv queues

sipa committed:
Handle mempool requests in send loop, subject to trickle

By eliminating queued entries from the mempool response and responding only at
trickle time, this makes the mempool no longer leak transaction arrival order
information (as the mempool itself is also sorted)-- at least no more than
relay itself leaks it.

gmaxwell commented on src/main.cpp:
```diff
@@ -5808,13 +5786,40 @@ bool SendMessages(CNode* pto)
+                    if (pto->minFeeFilter) {
+                        CFeeRate feeRate;
+                        mempool.lookupFeeRate(hash, feeRate);
+                        LOCK(pto->cs_feeFilter);
```
Perhaps the fee filter access should be hoisted out of this inner loop on transactions, and a local copy of the filtered value made.

sipa replied:
Done.

gmaxwell commented:
This is a very good idea-- much better than my earlier ones around avoiding mempool requests breaking privacy.

gmaxwell commented on src/main.cpp:
```diff
@@ -5808,13 +5786,40 @@ bool SendMessages(CNode* pto)
             }
+
+            // Determine transactions to relay
             if (fSendTrickle) {
```
"&& vInv.size() < INVENTORY_BROADCAST_MAX" perhaps? otherwise when a mempool request has prefilled the INV it will waste time heapifying for no reason.

sipa replied:
Done.

gmaxwell commented:
I have a test-harness now thats sutiable for testing this, a little 13 node network. Seems to work, it caught that the conversion to use the STL heap broke the behavior: the stl heap is a max heap.

gmaxwell commented:
Thanks for fixing, looking good so far.

We still get a _tiny_ number of orphans in a network the consists of nothing but this code: node A sends you Tx1. You getdata Tx1  Node B sends you Tx1 and Tx2 and Tx1 is a child of Tx1.  You getdata Tx2 from  node2.  Node2 responds faster. Tx2 is now an orphan.

sipa committed:
Return mempool queries in dependency order

sipa commented:
I added a commit to sort the response of mempool requests in dependency order, and renamed the PR to something more general.

dcousens commented:
concept ACK

sdaftuar commented on src/main.cpp:
```diff
@@ -5810,14 +5788,45 @@ bool SendMessages(CNode* pto)
-            if (fSendTrickle) {
+
+            // Determine transactions to relay
+            if (fSendTrickle && vInv.size() < INVENTORY_BROADCAST_MAX) {
```
This comparison on `vInv.size()` doesn't quite make sense, as we might have already filled it up and cleared it out above in the `mempool` command handling.  Most likely, if you've responded to a mempool command, you've used up your allotted bandwidth under this logic, but it's not clear to me whether those should really be considered together or not. 

With the change to `mempool` handling happening only when we do our next scheduled send, can we just restore the behavior where we respond with everything in the mempool, and then in those situations bail out on (ever) announcing the transactions that would otherwise have been announced?  Or is that too big a change to make to peers who might be treating responses to the mempool command differently somehow?  (That doesn't exactly make sense to me, as those aren't really distinguishable, but, I dunno...)

My suggestion would also result in a behavior change in the rare situation that a transaction to be relayed has been evicted from the mempool prior to announcement AND the peer has sent a mempool message in the interim, but at first glance that seems like an edge-case optimization to me, rather than an edge-case bug, as you'd probably choose not to relay something if you knew that you'd be evicting it in the very near future.

sdaftuar commented on src/main.cpp:
```diff
@@ -5779,49 +5770,93 @@ bool SendMessages(CNode* pto)
-                    pto->PushMessage(NetMsgType::INV, vInv);
-                    vInv.clear();
+            // Determine transactions to relay
+            if (fSendTrickle && vInv.size() < INVENTORY_BROADCAST_MAX) {
```
INVENTORY_BROADCAST_MAX is supposed to apply to rate-limited transaction relay, while vInv at this point can be holding blocks, and/or any leftovers from the mempool command.  This seems like it's not the right comparison.

sdaftuar commented on src/net.h:
```diff
@@ -396,14 +396,17 @@ class CNode
     CRollingBloomFilter filterInventoryKnown;
-    std::vector<CInv> vInventoryToSend;
+    std::set<uint256> vInventoryTxToSend;
+    std::vector<uint256> vInventoryBlockToSend;
```
nit: Perhaps add a comment explaining why the vector is needed for blocks?

sdaftuar commented:
Concept ACK, will test.

sdaftuar commented on src/main.h:
```diff
@@ -99,9 +99,12 @@ static const unsigned int MAX_REJECT_MESSAGE_LENGTH = 111;
+/** Average delay between trickled inventory transmissions in seconds.
+ *  Blocks and whitelisted receivers bypass this, outbound peers get half this delay. */
+static const unsigned int INVENTORY_BROADCAST_INTERVAL = 5;
+/** Maximum number of inventor items to send per transmission.
```
"inventor" -> "inventory"

sipa commented:
Added two commits to address the comments.

Mempool requests are now not filtered by vInventoryTxToSend, but instead every mempool inv sent as a response is removed from vInventoryTxToSend.

gmaxwell commented:
In a test topology over 48 hours this reduced the addition of orphan transactions from 436/hr to 2.3/hr.

gmaxwell commented:
Also, utack latest improvements.

dcousens commented:
concept ACK, utACK 98305db

gmaxwell commented:
@sipa: any squashing interest?

sipa commented:
Rebased and squashed.

sipa commented:
Also renamed the vInventoryTxToSend to setInventoryTxToSend, and restructured the tx send loop a bit to be less deeply nested.

gmaxwell commented:
Travis CI timeout, doesn't look like a real failure.

gmaxwell committed:
Move bloom and feerate filtering to just prior to tx sending.

This will avoid sending more pointless INVs around updates, and
 prevents using filter updates to timetag transactions.

Also adds locking for fRelayTxes.

sipa commented:
Added one more commit by @gmaxwell to move all inv filtering to the send stage (which avoids announcing things that are no longer in our mempool and bloom filter updates for those).

I'm sorry for the increasingly expansing scope this PR has had, so I'm not going to touch functionality any further unless bugs in it are found.

jonasnick commented on src/main.cpp:
```diff
@@ -5811,13 +5786,52 @@ bool SendMessages(CNode* pto)
+                    }
+                    pto->filterInventoryKnown.insert(hash);
+                    vInv.push_back(inv);
+                    if (vInv.size() == MAX_INV_SZ) {
```
It seems that in general it can happen that `vInv.size() > MAX_INV_SZ`, namely when `pto->vInventoryBlockToSend.size() >= MAX_INV_SZ`. Are we assuming that this isn't the case?

jonasnick replied:
not here https://github.com/sipa/bitcoin/blob/splitinvtxblock/src/main.cpp#L5790

sipa commented:
The entries to vInv are added one by one, and all such adds are followed by a max size test.

sipa commented on src/main.cpp:
```diff
@@ -5785,49 +5781,123 @@ bool SendMessages(CNode* pto)
+                    }
+                    pto->filterInventoryKnown.insert(hash);
+                    vInv.push_back(inv);
+                    if (vInv.size() == MAX_INV_SZ) {
```
Entries are added one by one to vInv, and all call sites where that happens are followed by a max size test, so I don't think vInv can exceed MAX_INV_SZ.

sipa replied:
Ah, good point. There may not be other mechanisms that prevent over 5000 blocks being inved at once.

sipa replied:
Fixed.

dcousens commented:
@sipa could you update your original PR description with the full scope of this PR now?

gmaxwell commented:
ACK.

jonasnick commented:
Ack

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -752,6 +752,31 @@ void CTxMemPool::check(const CCoinsViewCache *pcoins) const
+    DepthAndScoreComparator(CTxMemPool *mempool) : mp(mempool) {}
+    bool operator()(const uint256& a, const uint256& b) { return mp->CompareDepthAndScore(a, b); }
+};
+}
```
nit: What do you think of exposing the `DepthAndScoreComparator` defined here, and eliminating the `CompareInvMempoolOrder` object declared in `main.cpp` by moving the comparator it defines (which reverses the comparison) into this class?

sipa replied:
The comparator over there sorts iterators to set items, so there is more difference.

I do think the comparator (or at least a base comparator) should be exposed here, which grabs the mempool lock etc. Perhaps the other comparator can then take a ref to the first one as an argument to add the reversing and set iterator dereferencing.

sdaftuar replied:
I was thinking that because the main.cpp comparator acts on different objects, it'd be easy to just move that function into this class (and explain in comments what the two comparators were for and why they are subtly different!), but your suggestion seems reasonable to me as well.  I suppose its not awesome for the implementation details in main.cpp to leak into txmempool...

sipa replied:
Even better would be for txmempool.h to expose an opaque SortedMempoolSet
object, which takes a list of hashes as inputs, and maintains a sorted heap
of iterators to the mempool entries themselves, in addition to the cs lock.

sdaftuar commented:
ACK, apart from my question about whether we should combine the two comparators that are being introduced.

Couple notes:
 * `RelayTransaction` no longer requires a feerate, and `AcceptToMemoryPool` no longer needs to fill a feerate in.  We can clean this up in a future PR.
 * I noticed while testing that `CTxMemPool::queryHashes` is what we use in the RPC call `getrawmempool`.  It's hard for me to imagine anyone would care (and in fact the new sort order on the results seems handy!), but perhaps we should flag the change in the release notes?

jsantos4you commented on src/main.cpp:
```diff
@@ -4557,12 +4557,16 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             pfrom->cleanSubVer = SanitizeString(pfrom->strSubVer);
         }
-        if (!vRecv.empty())
+        if (!vRecv.empty()) {
```
3

gmaxwell commented:
I've had this running in its various forms for two weeks on a couple of public nodes, as well as a seperate test topology that consisted of nothing but nodes running this code. What do we need to move this forward?  I'm concerned that if I move to testing other PRs this will be forgotten.
